lua/lean/html.lua

```diff
@@ -0,0 +1,705 @@
+
+function Div:insert_new_tooltip(new_div)
+  local last_div = self.div_stack[#self.div_stack]
+  if last_div then
+    return last_div:add_tooltip(new_div)
+  else
+    return self:add_tooltip(new_div)
+  end
+end
+
+function Div:start_div(tags, text, name, hlgroup, listener)
+  local new_div = Div:new(tags, text, name, hlgroup, listener)
+  self:insert_new_div(new_div)
+  table.insert(self.div_stack, new_div)
+  return new_div
+end
+
+function Div:end_div()
+  table.remove(self.div_stack)
+end
+
+function Div:insert_div(tags, text, name, hlgroup, listener)
+  local new_div = self:start_div(tags, text, name, hlgroup, listener)
+  self:end_div()
+  return new_div
+end
+
+function Div:render()
+  local text = self.text
+  local hls = {}
+  for _, div in ipairs(self.divs) do
+    local new_text, new_hls = div:render()
+    for _, new_hl in ipairs(new_hls) do
+      new_hl.start = new_hl.start + #text
+      new_hl["end"] = new_hl["end"] + #text
+    end
+    vim.list_extend(hls, new_hls)
+    text = text .. new_text
+  end
+  local hlgroup = self.disabled and "htmlDivLoading" or self.hlgroup_override or self.hlgroup
+  if hlgroup then
+    if type(hlgroup) == "function" then
+      hlgroup = hlgroup(self)
+    end
+
+    if hlgroup then
+      table.insert(hls, {start = 1, ["end"] = #text, hlgroup = hlgroup, override = self.hlgroup_override})
+    end
+  end
+  self.hlgroup_override = nil
+  return text, hls
+end
+
+function Div:_pos_from_path(path)
+  if #path == 0 then return 1 end
+  path = {unpack(path)}
+
+  local this_branch = table.remove(path)
+  local this_name = this_branch.name
+  local this_idx = this_branch.idx
+
+  local pos = #self.text
+  for idx, child in ipairs(self.divs) do
+    if idx ~= this_idx then
+      pos = pos + #child:render()
+    else
+      if child.name ~= this_name then return nil end
+      local result = child:_pos_from_path(path)
+      return result and pos + result
+    end
+  end
+
+  return nil
+end
+
+function Div:pos_from_path(path)
+  path = {unpack(path)}
+
+  -- check that the first name matches
+  if self.name ~= table.remove(path).name then return nil end
+
+  return self:_pos_from_path(path)
+end
+
+function Div:_div_from_path(path, stack)
+  table.insert(stack, self)
+  if #path == 0 then return stack, self end
+  path = {unpack(path)}
+
+  local this_branch = table.remove(path)
+  local this_div = self.divs[this_branch.idx]
+  local this_name = this_branch.name
+
+  if not this_div or this_div.name ~= this_name then return nil, nil end
+
+  return this_div:_div_from_path(path, stack)
+end
+
+function Div:div_from_path(path)
+  path = {unpack(path)}
+
+  -- check that the first name matches
+  if self.name ~= table.remove(path).name then return nil, nil end
+
+  return self:_div_from_path(path, setmetatable({}, {__mode = "v"}))
+end
+
+function Div:_div_from_pos(pos, stack)
+  table.insert(stack, self)
+
+  local text = self.text
+
+  -- base case
+  if pos <= #text then return nil, stack, {} end
```

Comment:
When I click around randomly in the infoview (sorry, no better repro), I sometimes manage to get an exception from this line because `pos` can be nil here.

Reply:
I am using neovim-qt but I can also reproduce it in the terminal (with `set mouse=a`).

And now I've figured out how to reproduce it: you just need to select the final `--` with the mouse.

Reply:
I still can't reproduce it with the mouse. Are you still on 0.5? Perhaps an upstream issue is causing one of the failure states in `pos_to_raw_pos`?

Reply:
Ah I see, it's when you're doing a visual selection with the mouse, right? I think I've reproduced it.

Reply:
Yep, you can reproduce without the mouse as well when you do a visual selection and go past the end of any line.

Reply:
Should be fixed now.